### PR TITLE
pycbc_condition_strain: permit unscaled single precision output

### PR DESCRIPTION
**Pull request by titodalcanton**
It seems this would fix the problem @a-r-williamson is having with Virgo data in coh_PTF. What do people think?

### CODE REVIEW

**Comment by spxiwh** [comment]
I really don't like the current behaviour. It is confusing, which potentially causes problems.

I think that if the user specifies no option relating to DYN_RANGE_FAC then the data should *never* be resaled. To do otherwise leads to confusion (c.f. recent 2-hour delay while understanding this).

I would say that it's a good idea to read the data, rescale by DYN_RANGE_FAC, do conditioning, and then rescale back .... This avoids potential issues at single precision as all calculations are done after rescaling, but the output is *not* rescaled.

If there is a need for having rescaled output we can add that. But my main, and very strong, concern here is that the code *must not* rescale data if the user has not actually asked for that.

**Comment by a-r-williamson** [comment]
I agree with @spxiwh regarding scaling overall, but I also understand your reticence regarding changing the code incase it breaks anything. Is this code called from any workflow other than `pygrb`?

**Comment by ahnitz** [comment]
@titodalcanton Is the idea here to allow single precision to work without rescaling the output? If that's the case, we should have an option to *undo* the rescaling so the output is not affected by the dynamic range factor.

**Comment by spxiwh** [comment]
The previous behaviour (previous to #1838, which fixed some genuine bugs) was that double-precision data would not be rescaled. You will probably find that #1838 has broken the GRB code if you're using `pycbc_condition_strain` in a workflow.

... Given this, and that the previous behaviour was *different* for single and double precision, I think that backwards compatibility issues cannot be avoided here, so I think we should do the most obvious thing doing forward.

**Comment by titodalcanton** [comment]
It is true that the purpose and behavior of this program evolved enough over time that we may as well rework the CLI to a saner version. How does this look? I do not think there is any other use in the workflow other than PyGRB, however I suspect people have scripts lying around which rely on it (including CW people at the AEI). Hopefully I documented this version well enough and hopefully those people will instinctively read the docs when things break ;)

File: bin/pycbc_condition_strain [hunk 1]
```diff
@@ -45,28 +44,36 @@ parser.add_argument('--output-channel-name',
...
-                 'dynamic range factor, as it would lead to underflows')
-precision = 'single' if args.single_precision else 'double'
-dyn_range_factor = 1 if args.no_dyn_range_factor else pycbc.DYN_RANGE_FAC
-out_strain = pycbc.strain.from_cli(args, dyn_range_factor, precision=precision)
+# read and condition strain as pycbc_inspiral would do
+out_strain = pycbc.strain.from_cli(args, dyn_range_fac=pycbc.DYN_RANGE_FAC)
+
+if args.output_precision == 'single' and not args.dyn_range_factor:
+    logging.warn('Warning: saving strain in single precision without the '
+                 'dynamic range scaling factor. This might lead to underflows')
```
Comment (spxiwh) [inline bin/pycbc_condition_strain]: I think this warning is not needed. It only applies if squaring the data during conditioning:

https://en.wikipedia.org/wiki/IEEE_754-1985

states that we are not approaching float limits for the raw strain any time soon!

**Comment by spxiwh** [comment]
I'm basically happy to merge this, and it should allow the GRB use-case. I wonder if the `dyn-scale-factor` should accept a float so that the user can specify what to rescale by ... But then I wonder if it's needed at all! Anyway, these are minor points, so do whatever you're happy with, and merge after you are happy this is tested.

**Comment by spxiwh** [comment]
+1 For ensuring that any compatibility issues cause a code failure so that people need to read the help message!

**Comment by a-r-williamson** [comment]
This is a nice way to do this, but actually I think line 68 leads to an error:
```
2017-08-29 03:29:43,764 Writing output gates
Traceback (most recent call last):
  File "/home/andrew.williamson/pycbc-pygrb-env/bin/pycbc_condition_strain", line 6, in <module>
    exec(compile(open(__file__).read(), __file__, 'exec'))
  File "/mnt/qfs3/andrew.williamson/pycbc-pygrb-env/src/pycbc/bin/pycbc_condition_strain", line 86, in <module>
    for k, v in out_strain.gating_info.items():
AttributeError: 'TimeSeries' object has no attribute 'gating_info'
```

File: bin/pycbc_condition_strain [hunk 2]
```diff
@@ -45,28 +45,32 @@ parser.add_argument('--output-channel-name',
...
-                 'dynamic range factor, as it would lead to underflows')
-precision = 'single' if args.single_precision else 'double'
-dyn_range_factor = 1 if args.no_dyn_range_factor else pycbc.DYN_RANGE_FAC
-out_strain = pycbc.strain.from_cli(args, dyn_range_factor, precision=precision)
+# read and condition strain as pycbc_inspiral would do
+out_strain = pycbc.strain.from_cli(args, dyn_range_fac=pycbc.DYN_RANGE_FAC)
+
+# force strain precision to be as requested
+out_strain = out_strain.astype(
+        float32 if args.output_precision == 'single' else float64)
```
Comment (spxiwh) [inline bin/pycbc_condition_strain]: To fix the error @a-r-williamson sees, why not just remove the two lines above, and send the precision key-word argument into from_cli?

Comment (titodalcanton) [reply bin/pycbc_condition_strain]: But that was exactly the problem we were having before no?

Comment (spxiwh) [reply bin/pycbc_condition_strain]: Now I'm confused! What problem are you referring to. This type conversion would be done within the strain module, no?

https://github.com/ligo-cbc/pycbc/blob/master/pycbc/strain.py#L274

Comment (titodalcanton) [reply bin/pycbc_condition_strain]: The precision argument of `from_cli` influences the internal precision of how autogating is done, which I do not like. What I have done now exactly matches what `pycbc_inspiral` does, I think that is the way it should behave. @a-r-williamson 's error can be fixed trivially by moving a block of code, which I just did. I will upload in a moment, after a quick test.

Comment (spxiwh) [reply bin/pycbc_condition_strain]: Okay, this suggests that we need to do some cleaning of `from_cli`, but that can wait until O2 is properly over.

Comment (spxiwh) [reply bin/pycbc_condition_strain]: Patch remains approved when you want to merge

**Comment by titodalcanton** [comment]
@a-r-williamson can you try this again?

**Comment by a-r-williamson** [comment]
@titodalcanton thanks, that looks great.

**Pull request merged**